tfx/examples/cifar10/cifar10_utils_native_keras.py

```diff
@@ -0,0 +1,259 @@
+from tfx.components.trainer.rewriting import rewriter
+from tfx.components.trainer.rewriting import rewriter_factory
+
+from tfx.components.trainer.executor import TrainerFnArgs
+
+# CIFAR10 dataset has 50000 train records, and 10000 test records
+_TRAIN_DATA_SIZE = 50000
+_EVAL_DATA_SIZE = 10000
+_TRAIN_BATCH_SIZE = 64
+_EVAL_BATCH_SIZE = 64
+
+IMAGE_KEY = 'image'
+LABEL_KEY = 'label'
+
+def transformed_name(key):
+  return key + '_xf'
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _get_serve_tf_examples_fn(model):
+  """Returns a function that feeds the input tensor into the model."""
+
+  @tf.function
+  def serve_tf_examples_fn(image_tensor):
+    """Returns the output to be used in the serving signature."""
+    return model(image_tensor)
+
+  return serve_tf_examples_fn
+
+def _input_fn(file_pattern: List[Text],
+              tf_transform_output: tft.TFTransformOutput,
+              batch_size: int = 200) -> tf.data.Dataset:
+  """Generates features and label for tuning/training.
+
+  Args:
+    file_pattern: List of paths or patterns of input tfrecord files.
+    tf_transform_output: A TFTransformOutput.
+    batch_size: representing the number of consecutive elements of returned
+      dataset to combine in a single batch
+
+  Returns:
+    A dataset that contains (features, indices) tuple where features is a
+      dictionary of Tensors, and indices is a single Tensor of label indices.
+  """
+  transformed_feature_spec = (
+      tf_transform_output.transformed_feature_spec().copy())
+
+  dataset = tf.data.experimental.make_batched_features_dataset(
+      file_pattern=file_pattern,
+      batch_size=batch_size,
+      features=transformed_feature_spec,
+      reader=_gzip_reader_fn,
+      label_key=transformed_name(LABEL_KEY))
+
+  return dataset
+
+def _freeze_model_by_percentage(model: tf.keras.Model,
+                                percentage: float):
+  """Freeze part of the model based on specified percentage
+
+  Args:
+    model: The keras model need to be partially frozen
+    percentage: the percentage of layers to freeze
+  """
+  if percentage < 0 or percentage > 1:
+    raise Exception("freeze percentage should between 0.0 and 1.0")
+
+  num_layers = len(model.layers)
+  num_layers_to_freeze = int(num_layers * percentage)
+  for idx, layer in enumerate(model.layers):
+    if idx < num_layers_to_freeze:
+      layer.trainable = False
+    else:
+      layer.trainable = True
+
+def _build_keras_model() -> tf.keras.Model:
+  """Creates a MobileNet model pretrained on ImageNet for finetuning
+
+  Args:
+    model: The keras model need to be partially frozen
+  Returns:
+    A Keras Model.
+  """
+  base_model = tf.keras.applications.MobileNet(
+      input_shape=(224, 224, 3), include_top=False, weights='imagenet',
+      pooling='avg')
+
+  model = tf.keras.Sequential([
```

Comment:
Sure I can check that

Reply:
4 labels in cifar-10 do not exist in ImageNet, and 6 labels "some what" exist. The labels in ImageNet are fine-grained categories, for example, there are 5 labels in ImageNet that are related to cat: "tabby cat", "tiger cat", "Persian cat", "Siamese cat", "Egyptian cat", whereas in cifar-10 there is only one label named "cat".

Reply:
ok cool, can we check how our model does on those cifar-10 labels?

Reply:
hmmm i feel there is no direct way to do this? The origin pre-trained model has Dense(1000) as the final classifier, and each output corresponds to one ImageNet class. Since there is no one-to-one mapping between these 1000 classes and cifar-10 class, i think we could not test its performance for cifar-10?